### PR TITLE
feat: LT-1639 - Edit goods/counties matrix on major edits

### PR DESCRIPTION
**Pull request by jfaracik**
* Added audit for assigning goods to countries
* Removed goods_type_pk from view as it isn't used
* Assigning goods to countries now requires you to select at least one country

### CODE REVIEW

File: applications/views/goods.py [hunk 1]
```diff
@@ -212,15 +212,34 @@ class ApplicationGoodsTypeCountries(APIView):
...
             if not Country.objects.filter(pk__in=countries).count() == len(countries):
                 return HttpResponse(status=status.HTTP_404_NOT_FOUND)
 
+            initial_countries = list(good.countries.all())
             good.countries.set(countries)
+            updated_countries = list(good.countries.all())
+
+            if initial_countries != updated_countries:
+                # Assigning the good to no countries is the same as assigning the good to all countries
+                if not updated_countries:
```
[the remark this answers is not in the document]
Comment (efahy) [reply applications/views/goods.py]: Looking at the post for goodstypes you will need to include a similar validation there, to make it consistent

File: applications/views/goods.py [hunk 2]
```diff
@@ -212,15 +212,43 @@ class ApplicationGoodsTypeCountries(APIView):
...
+            added_countries = [country.name for country in good.countries.all() if country not in initial_countries]
+
+            if removed_countries:
+                audit_trail_service.create(
+                    actor=request.user,
+                    verb=AuditType.REMOVED_COUNTRIES_FROM_GOOD,
+                    action_object=good,
+                    target=Case.objects.get(id=application.id),
+                    payload={"good_type_name": good.description, "countries": ", ".join(removed_countries),},
+                )
```
Comment (efahy) [inline applications/views/goods.py]: Please add API test that this audit is created when appropriate

File: applications/views/goods.py [hunk 3]
```diff
@@ -212,15 +212,43 @@ class ApplicationGoodsTypeCountries(APIView):
...
+
+            if added_countries:
+                audit_trail_service.create(
+                    actor=request.user,
+                    verb=AuditType.ASSIGNED_COUNTRIES_TO_GOOD,
+                    action_object=good,
+                    target=Case.objects.get(id=application.id),
+                    payload={"good_type_name": good.description, "countries": ", ".join(added_countries),},
+                )
```
Comment (efahy) [inline applications/views/goods.py]: API test for this audit entry please

File: applications/views/goods.py [hunk 4]
```diff
@@ -212,15 +212,43 @@ class ApplicationGoodsTypeCountries(APIView):
...
-    def put(self, request, application, goodstype_pk):
+    def put(self, request, application):
         data = request.data
 
         for good, countries in data.items():
             good = get_goods_type(good)
+
+            # Validate that at least one country has been selected per good
+            if not countries:
+                return JsonResponse({"errors": "Select at least one country for each good"})
```
Comment (efahy) [inline applications/views/goods.py]: API test for this condition please

File: applications/views/goods.py [hunk 5]
```diff
@@ -212,15 +212,43 @@ class ApplicationGoodsTypeCountries(APIView):
...
+
+            # Validate that the countries given are valid countries
             if not Country.objects.filter(pk__in=countries).count() == len(countries):
                 return HttpResponse(status=status.HTTP_404_NOT_FOUND)
 
+            initial_countries = list(good.countries.all())
             good.countries.set(countries)
+            removed_countries = [country.name for country in initial_countries if country not in good.countries.all()]
+            added_countries = [country.name for country in good.countries.all() if country not in initial_countries]
```
Comment (efahy) [inline applications/views/goods.py]: Add some API tests to ensure that the put adds and removes countries as expected

File: applications/tests/tests_managing_countries_on_goods_type.py [hunk 6]
```diff
@@ -73,6 +72,19 @@ def test_state_can_be_over_written(self):
         self.assertTrue(self.country_1 in self.goods_type_1.countries.all())
         self.assertTrue(self.country_2 in self.goods_type_1.countries.all())
 
+    def test_cannot_set_no_countries_on_good(self):
```
Comment (cstobart) [inline applications/tests/tests_managing_countries_on_goods_type.py]: Does this make the goods location section no longer optional? Is there still a way to assign all countries to all goods by skipping the sections?

File: applications/views/goods.py [hunk 7]
```diff
@@ -212,15 +212,45 @@ class ApplicationGoodsTypeCountries(APIView):
...
+    def put(self, request, application):
         data = request.data
 
         for good, countries in data.items():
             good = get_goods_type(good)
+
+            # Validate that at least one country has been selected per good
+            if not countries:
+                return JsonResponse(
+                    {"errors": "Select at least one country for each good"}, status=status.HTTP_400_BAD_REQUEST
```
Comment (cstobart) [inline applications/views/goods.py]: move string into a content file

File: audit_trail/payload.py [hunk 8]
```diff
@@ -32,6 +32,8 @@ class AuditType(Enum):
     UPDATE_APPLICATION_LETTER_REFERENCE = "updated the letter reference from " "{old_ref_number} to {new_ref_number}"
     ADDED_APPLICATION_LETTER_REFERENCE = "added the letter reference: {new_ref_number}"
     REMOVED_APPLICATION_LETTER_REFERENCE = "removed the letter reference: {old_ref_number}"
+    ASSIGNED_COUNTRIES_TO_GOOD = "added the destinations {countries} to '{good_type_name}'"
```
Comment (cstobart) [inline audit_trail/payload.py]: Might it be worth visually separating Exporter audit items from Gov audit items?

File: conf/exceptions.py [hunk 9]
```diff
@@ -45,6 +45,21 @@ def __init__(self, detail=None, code=None):
         self.detail = _get_error_details({"errors": detail}, code)
 
 
+class PermissionDeniedError(APIException):
+    status_code = status.HTTP_403_FORBIDDEN
+    default_detail = "You don't have permission to do that action"
```
Comment (cstobart) [inline conf/exceptions.py]: Put this into a content file unless it is inaccessible

File: goodstype/serializers.py [hunk 10]
```diff
@@ -49,6 +46,12 @@ def __init__(self, *args, **kwargs):
             if hasattr(self, "initial_data"):
                 self.initial_data["control_code"] = None
 
+    def get_countries(self, instance):
+        countries = instance.countries
+        if not countries.count():
```
Comment (cstobart) [inline goodstype/serializers.py]: Is `.count` required?

File: applications/views/goods.py (hunk 4 above)
Comment (jfaracik) [reply applications/views/goods.py]: Done!

File: applications/tests/tests_managing_countries_on_goods_type.py (hunk 6 above)
Comment (jfaracik) [reply applications/tests/tests_managing_countries_on_goods_type.py]: Yep! All countries are selected by default now after a discussion with Eugene. We weren't a fan of how all countries selected == no countries selected - it felt odd, this also simplifies our audit entry.

File: goodstype/serializers.py (hunk 10 above)
Comment (jfaracik) [reply goodstype/serializers.py]: Yeah it returns something like object.country.Country otherwise :/

File: conf/exceptions.py (hunk 9 above)
Comment (jfaracik) [reply conf/exceptions.py]: It's just for devs :)

File: applications/tests/tests_managing_countries_on_goods_type.py (hunk 6 above)
Comment (cstobart) [reply applications/tests/tests_managing_countries_on_goods_type.py]: 👍

File: conf/exceptions.py (hunk 9 above)
Comment (cstobart) [reply conf/exceptions.py]: 👍

File: goodstype/serializers.py (hunk 10 above)
Comment (cstobart) [reply goodstype/serializers.py]: Yeah that's fair

**Pull request merged**